First cut of text-to-speech tutorial (#6783)
godotengine/godot-docs: tutorials/audio/text_to_speech.rst

.. _doc_text_to_speech:

> Text to Speech
> ==============

Reviewer: ```suggestion
Text to speech
==============
```

Requirements for functionality
------------------------------

Godot includes text-to-speech functionality. You can find these under the :ref:`DisplayServer class <class_DisplayServer>`.

> Godot depends on system libraries for text-to-speech functionality. These libraries are installed by default on Windows and MacOS, but not on all Linux distros. If they are not present, text-to-speech functionality will not work. Speifically, the ``tts_get_voices`` method will return an empty list, indicating that there are no usable voices.

Reviewer 2: ```suggestion
Godot depends on system libraries for text-to-speech functionality. These libraries are installed by default on Windows and MacOS, but not on all Linux distributions. If they are not present, text-to-speech functionality will not work. Speifically, the ``tts_get_voices`` method will return an empty list, indicating that there are no usable voices.
```

> Godot includes text-to-speech functionality. You can find these under the :ref:`DisplayServer class <class_DisplayServer>`.
> 
> Godot depends on system libraries for text-to-speech functionality. These libraries are installed by default on Windows and MacOS, but not on all Linux distros. If they are not present, text-to-speech functionality will not work. Speifically, the ``tts_get_voices`` method will return an empty list, indicating that there are no usable voices.
> 
> Both Godot users on Linux and end-users on Linux running Godot games need to ensure that their system includes the system libraries for text-to-speech to work. Please consult with your own distro's documentation to determine what libraries you need to install.
> 
> For example, on Arch-based systems, Godot requires ``spd-say``, ``festival``, and ``espeakup``.

Reviewer: ```suggestion
Godot includes text-to-speech functionality. You can find these under the :ref:`DisplayServer class <class_DisplayServer>`.

Godot depends on system libraries for text-to-speech functionality. These libraries are installed by default on Windows and macOS, but not on all Linux distributions. If they are not present, text-to-speech functionality will not work. Specifically, the ``tts_get_voices()`` method will return an empty list, indicating that there are no usable voices.

Both Godot users on Linux and end-users on Linux running Godot games need to ensure that their system includes the system libraries for text-to-speech to work. Please consult with your own distibution's documentation to determine what libraries you need to install.

For example, on Arch Linux-based systems, Godot requires ``spd-say``, ``festival``, and ``espeakup``.
```

> Best Practices

Reviewer: ```suggestion
Best practices
```

.. _doc_text_to_speech:

Text to speech
==============

Basic Usage
-----------

Basic usage of text-to-speech involves the following one-time steps:

- Query the system for a list of usable voices
- Store the ID of the voice you want to use

Once you have the voice ID, you can use it to speak some text:

::

    # One-time steps
    var voices:Array = DisplayServer.tts_get_voices()
    # Pick a voice. Here, we arbitrarily pick the first voice.

>     var voice_data:Dictionary = voices[0]

Author: While this is true on Linux, when I tried this on Windows, I got two voices, and I don't recall that either had an indication of the language (English?).

On Linux, I got ~13k languages.

>     var voice_data:Dictionary = voices[0]

Reviewer 3: All voices should have "language", "name", and "id" (it might be all English, if nothing else is installed). If "language" is empty, there might be a bug. It's definitely returning correct languages (`xx_XX` codes) on my Windows pc.

    # Say "Hello, World!"
    DisplayServer.tts_speak("Hello, World!", voice_data["id"])

    # Say a longer sentence, and then interrupt it.
    # Note that this method is asynchronous: execution proceeds
    # to the next line immediately, before the voice finishes speaking.
    
    var long_message:String = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur"

    DisplayServer.tts_speak(long_message)

    # Immediately stop the current text mid-sentence and say goodbye instead
    DisplayServer.tts_stop()
    DisplayServer.tts_speak("Goodbye!")

Requirements for functionality
------------------------------

Godot includes text-to-speech functionality. You can find these under the :ref:`DisplayServer class <class_DisplayServer>`.

> Godot depends on system libraries for text-to-speech functionality. These libraries are installed by default on Windows and macOS, but not on all Linux distributions. If they are not present, text-to-speech functionality will not work. Specifically, the ``tts_get_voices()`` method will return an empty list, indicating that there are no usable voices.

Reviewer 3: ```suggestion
Godot depends on system libraries for text-to-speech functionality. These libraries are installed by default on Android, Windows, iOS and macOS, but not on all Linux distributions. If they are not present, text-to-speech functionality will not work. Specifically, the ``tts_get_voices()`` method will return an empty list, indicating that there are no usable voices.
```

Availability for the Web exports depends on OS.

Both Godot users on Linux and end-users on Linux running Godot games need to ensure that their system includes the system libraries for text-to-speech to work. Please consult with your own distibution's documentation to determine what libraries you need to install.

For example, on Arch Linux-based systems, Godot requires ``spd-say``, ``festival``, and ``espeakup``.

Best practices
--------------

> The best practices for text-to-speech, in terms of the ideal player experience for blind players, is to integrate directly with the screen-reader APIs (such as NVDA's API) to send output directly to it. This provides advanced functionality for blind players, such as being able to traverse previous output forwards and backwards, to replay, and to play at a faster or slower rate as suits them. Godot currently doesn't support screen-reader API integration.

Author: Have you seen the godot_tts repo? https://github.com/lightsoutgames/godot-tts

My understanding is that it includes a couple of DLLs or SO files, and is able to interface directly with NVDA. I assume this is through some sort of wrapper around a C++ call or something similar (from when I discussed it with the repo API). That was what I was referring to; is there a better way to describe it than "NVDA API"?

As for tts_speak having arguments, my point is not that (I mention later on allowing users to specify, and persist, values for control rate pitch etc.) - my point is simply that these are already pre-configured to the user's preferences as part of NVDA's configuration. Is there a better way to word this?

> The best practices for text-to-speech, in terms of the ideal player experience for blind players, is to integrate directly with the screen-reader APIs (such as NVDA's API) to send output directly to it. This provides advanced functionality for blind players, such as being able to traverse previous output forwards and backwards, to replay, and to play at a faster or slower rate as suits them. Godot currently doesn't support screen-reader API integration.

Reviewer 3: Apparently, NVDA have [direct TTS API](https://github.com/nvaccess/nvda/tree/master/extras/controllerClient), but it's not a screen-reader API (basically provides only `speak(text)` and `stop` functions), so it's definitely not a best practice and not a proper screen reader support. Also, it's wrong to assume NVDA is used (not user what this substitution dll will do if it's not installed).

I assume `SAAPI64.dll` replacement is intended for the end users to substitute SAPI TTS output with the NVDA output, not for developers.

With the current state of the Godot text-to-speech APIs, best practices include:

- Develop the game with text-to-speech enabled, and ensure that everything sounds correct
- Allow players to control which voice to use, and save/persist that selection across game sessions
- Allow players to control the speech rate, and save/persist that selection across game sessions

This provides your blind players with the most flexibility and comfort available when not using a screen reader, and minimizes the chance of frustrating and alienating them.

Caveats and Other Information
-----------------------------

> - Expect delays when you call `tts_speak` and `tts_stop`. The actual delay time varies depending on both the OS and on your machine's specifications.

Reviewer 3: This is especially relevant for Android and Web, since some of the voices are online only, and therefore depends on server load, network latency and speed.

> - Non-English text doesn't seem to be supported; even on Linux, with language-specific voices such as Arabic, the text reads out letter by letter.

Reviewer 3: See the https://www.ghacks.net/2018/08/11/unlock-all-windows-10-tts-voices-system-wide-to-get-more-of-them/ in particular, with this registry manipulations Arabic works on Windows as well (tested on Windows 11 with Hoda voice).

> - Non-English text doesn't seem to be supported; even on Linux, with language-specific voices such as Arabic, the text reads out letter by letter.

Author: > with this registry manipulations Arabic works on Windows as well

I don't think it's a good idea to add this, because end-users will need to make the same change. From my understanding, based on the blind players I spoke to, most TTS users use English.

As for Arabic, when I selected an Arabic voice on Linux and used tts_speak with some Arabic text, it came out garbled. Maybe I should just remove this line altogether? I'm not sure why it worked for you on Windows and not for me on Linux (maybe something to do with the underlying system APIs).

- Non-ASCII characters, such as umlaut, are similarly not supported (e.g. ö reads as "o umlaut")
- Most blind players also use Windows with the NVDA screen reader.
- Windows text-to-speech APIs generally perform better than their equivalents on other systems (e.g. `tts_stop` followed by `tts_speak` immediately speaks the new message).

> - Some systems, such as Linux, provide several voices, including voices with different accents and for different languages.

Reviewer 3: All systems do, but usually only voice for the current OS locale is installed by default.

+------------------+-----------------------------------------------------------------------------------------------------------+

Best practices
--------------

The best practices for text-to-speech, in terms of the ideal player experience for blind players, is to send output to the player's screen reader. This preserves the choice of language, speed, pitch, etc. that the user set, as well as allows advanced features like allowing players to scroll backward and forward through text. As of now, Godot doesn't provide this level of integration.

With the current state of the Godot text-to-speech APIs, best practices include:

- Develop the game with text-to-speech enabled, and ensure that everything sounds correct
- Allow players to control which voice to use, and save/persist that selection across game sessions
- Allow players to control the speech rate, and save/persist that selection across game sessions

This provides your blind players with the most flexibility and comfort available when not using a screen reader, and minimizes the chance of frustrating and alienating them.

Caveats and Other Information
-----------------------------

> - Expect delays when you call `tts_speak` and `tts_stop`. The actual delay time varies depending on both the OS and on your machine's specifications. This is especially critical on Android and Web, where some of the voices depend on web-services, and the actual time to playback depends on server load, network latency, and other factors.

Reviewer 2: ```suggestion
- Expect delays when you call `tts_speak` and `tts_stop`. The actual delay time varies depending on both the OS and on your machine's specifications. This is especially critical on Android and Web, where some of the voices depend on web services, and the actual time to playback depends on server load, network latency, and other factors.
```
